File: qt/gen_qerrormessage.cpp
// +build ignore

#include <QDialog>
#include <QErrorMessage>
#include <QEvent>
#include <QMetaObject>
#include <QObject>
#include <QPaintDevice>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWidget>
#include <qerrormessage.h>
#include "gen_qerrormessage.h"
class MiqtVirtualQErrorMessage : public virtual QErrorMessage {
public:
MiqtVirtualQErrorMessage(QWidget* parent): QErrorMessage(parent) {};
MiqtVirtualQErrorMessage(): QErrorMessage() {};

virtual ~MiqtVirtualQErrorMessage() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QErrorMessage::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QErrorMessage_MetaObject(const_cast<MiqtVirtualQErrorMessage*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QErrorMessage::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QErrorMessage::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QErrorMessage_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QErrorMessage::qt_metacast(param1);

	}
};
QErrorMessage* QErrorMessage_new(QWidget* parent) {
return new MiqtVirtualQErrorMessage(parent);
}
QErrorMessage* QErrorMessage_new2() {
return new MiqtVirtualQErrorMessage();
}
void QErrorMessage_virtbase(QErrorMessage* src
, QDialog** outptr_QDialog
) {
*outptr_QDialog = static_cast<QDialog*>(src);
}
QMetaObject* QErrorMessage_MetaObject(const QErrorMessage* self) {
	return (QMetaObject*) self->metaObject();
}
void* QErrorMessage_Metacast(QErrorMessage* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QErrorMessage_Tr(const char* s) {
	QString _ret = QErrorMessage::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QErrorMessage* QErrorMessage_QtHandler() {
	return QErrorMessage::qtHandler();
}
void QErrorMessage_ShowMessage(QErrorMessage* self, struct miqt_string message) {
	QString message_QString = QString::fromUtf8(message.data, message.len);
	self->showMessage(message_QString);
}
void QErrorMessage_ShowMessage2(QErrorMessage* self, struct miqt_string message, struct miqt_string typeVal) {
	QString message_QString = QString::fromUtf8(message.data, message.len);
	QString typeVal_QString = QString::fromUtf8(typeVal.data, typeVal.len);
	self->showMessage(message_QString, typeVal_QString);
}
struct miqt_string QErrorMessage_Tr2(const char* s, const char* c) {
	QString _ret = QErrorMessage::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QErrorMessage_Tr3(const char* s, const char* c, int n) {
	QString _ret = QErrorMessage::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QErrorMessage_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQErrorMessage*>( (QErrorMessage*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QErrorMessage_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQErrorMessage*)(self) )->virtualbase_MetaObject();
}
void QErrorMessage_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQErrorMessage*>( (QErrorMessage*)(self) )->handle__Metacast = slot;
}
void* QErrorMessage_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQErrorMessage*)(self) )->virtualbase_Metacast(param1);
}
void QErrorMessage_Delete(QErrorMessage* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQErrorMessage*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qerrormessage.go
package qt

import (
	"unsafe"
)

type QErrorMessage struct {
	h          uintptr
	isSubclass bool
}

// NewQErrorMessage constructs a new QErrorMessage object.
func NewQErrorMessage(parent *QWidget) *QErrorMessage {
	ret := newQErrorMessage(QErrorMessage_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQErrorMessage2 constructs a new QErrorMessage object.
func NewQErrorMessage2() *QErrorMessage {
	ret := newQErrorMessage(QErrorMessage_new2())
	ret.isSubclass = true
	return ret
}

func (this *QErrorMessage) MetaObject() *QMetaObject {
	return newQMetaObject(QErrorMessage_MetaObject(this.h))
}

func (this *QErrorMessage) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QErrorMessage_Metacast(this.h, param1_Cstring))
}

func QErrorMessage_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QErrorMessage_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QErrorMessage_QtHandler() *QErrorMessage {
	return newQErrorMessage(QErrorMessage_QtHandler())
}

func (this *QErrorMessage) ShowMessage(message string) {
	message_ms := struct_miqt_string{}
	message_ms.data = CString(message)
	message_ms.len = size_t(len(message))
	defer free(unsafe.Pointer(message_ms.data))
	QErrorMessage_ShowMessage(this.h, message_ms)
}

func (this *QErrorMessage) ShowMessage2(message string, typeVal string) {
	message_ms := struct_miqt_string{}
	message_ms.data = CString(message)
	message_ms.len = size_t(len(message))
	defer free(unsafe.Pointer(message_ms.data))
	typeVal_ms := struct_miqt_string{}
	typeVal_ms.data = CString(typeVal)
	typeVal_ms.len = size_t(len(typeVal))
	defer free(unsafe.Pointer(typeVal_ms.data))
	QErrorMessage_ShowMessage2(this.h, message_ms, typeVal_ms)
}

func QErrorMessage_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QErrorMessage_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QErrorMessage_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QErrorMessage_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QErrorMessage) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QErrorMessage_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QErrorMessage) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QErrorMessage_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QErrorMessage_MetaObject
func miqt_exec_callback_QErrorMessage_MetaObject(self QErrorMessage, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QErrorMessage{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QErrorMessage) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QErrorMessage_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QErrorMessage) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QErrorMessage_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QErrorMessage_Metacast
func miqt_exec_callback_QErrorMessage_Metacast(self QErrorMessage, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QErrorMessage{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
